quill-core/src/main/scala/io/getquill/quotation/Parsing.scala

```diff
@@ -438,13 +438,31 @@ trait Parsing {
   private def is[T](tree: Tree)(implicit t: TypeTag[T]) =
     tree.tpe <:< t.tpe
 
+  private def isCaseClass[T: WeakTypeTag] = {
+    val symbol = c.weakTypeTag[T].tpe.typeSymbol
+    symbol.isClass && symbol.asClass.isCaseClass
+  }
+
+  private def firstConstructorParamList[T: WeakTypeTag] = {
+    val tpe = c.weakTypeTag[T].tpe
+    val paramLists = tpe.decls.collect {
+      case m: MethodSymbol if m.isConstructor => m.paramLists.map(_.map(_.name))
+    }
+    paramLists.toList(0)(0).map(_.toString)
+  }
+
   val valueParser: Parser[Ast] = Parser[Ast] {
-    case q"null" => NullValue
+    case q"null"                         => NullValue
     case Literal(c.universe.Constant(v)) => Constant(v)
-    case q"((..$v))" if (v.size > 1) => Tuple(v.map(astParser(_)))
+    case q"((..$v))" if (v.size > 1)     => Tuple(v.map(astParser(_)))
+    case q"new $ccTerm(..$v)" if (isCaseClass(c.WeakTypeTag(ccTerm.tpe.erasure))) => {
```

Comment:
I think it'd be possible to support case classes without `new` if you use a pattern like:
`q"$ccTem.apply[$..t](..$v)"`

Reply:
So I tried '.apply' in the quasi quotes but then two things happened:
1. Only the Person.apply(...) case was matched, not the Person(...)
2. The tpe.typeSymbol.isClass was false or tpe was null depending on the scenario.

What I haven't figured out is how to check whether $ccTerm symbol is a companion of some given case class. Any ideas?

Reply:
> Only the Person.apply(...) case was matched, not the Person(...)

Yeah, you'd need to have both patterns



> The tpe.typeSymbol.isClass was false or tpe was null depending on the scenario.
> What I haven't figured out is how to check whether $ccTerm symbol is a companion of some given case class. Any ideas?

I think you could use `isModule`

Reply:
Wait.. you mean `q"$ccTem.apply[..$t](..$v)"` right? ($ was before the dots which confused me). I tried doing that and it matches but it says $t is an empty list :(

Reply:
yeah, it was a typo. Yes, it'll be an empty list if the case class doesn't have a type parameter. I think you can just ignore `t`. Is there an error message?

Reply:
So I figured out an actual way to do this but there are two caveats:
1. The code looks awful
2. It doesn't work with case classes defined inside a method. They have to be either top-level or inside another object. That's the only way the `.companion` method seems to work.

Some more elaboration of point 1 as follows:
Since just about everything in the world matches the `q"$ccCompanion(..$v)"` pattern I have to do a bunch of secondary checks.
1. Check that ccCompanion.apply is a type constructor.
2. Check that the ccCompanion.apply returns a case class.
2. Check that ccCompanion's 1st child is a module.
3. Check that this module's companion symbol is the same as the case class found in Step 2.

It looks something like this:
````scala
  private def ifThenSome[T](cond: =>Boolean, output : =>T):Option[T] =
    if (cond) Some(output) else None

  private def isSyntheticCaseClassCompanion(ccCompanion:Tree):Boolean = {
    val output = for {
      resultType <- ifThenSome(
        isTypeConstructor(c.WeakTypeTag(ccCompanion.tpe.erasure)),
        resultType(c.WeakTypeTag(ccCompanion.tpe.erasure))
      )
      firstChild <- ifThenSome(
        isCaseClass(c.WeakTypeTag(resultType)) && ccCompanion.children.length > 0,
        ccCompanion.children(0)
      )
      moduleClass <- ifThenSome(
        isSyntheticModuleClass(c.WeakTypeTag(firstChild.tpe.erasure)),
        asClass(c.WeakTypeTag(firstChild.tpe.erasure))
      )
    } yield (moduleClass.companion == resultType.typeSymbol)
    output.getOrElse(false)
  }
````
This seems to work in most cases except for the case where the case class is defined inside of a method body. In that situation, the `.companion` method returns `NoSymbol`.
What do you think?

Reply:
Thanks for digging into it! The macro API is pretty complicated :( I can't test it now, but I wonder if something like this would work:

```scala
case tree @ q"$ccCompanion.apply[..$t](..$v)" if t.tpe.companion == ccCompanion && isCaseClass(t.tpe)
```

I think can check out your branch tomorrow and try it out if you like. Also, I think it's important to have `[..$t]` so it'll also support case classes with type parameters.

Reply:
I’m totally confused now. How is `t.tpe.companion` possible? In most of the
experiments I’ve done, `$t` is a Nil list. Would definitely appreciate it if you could take a closer look.

Reply:
BTW, I think that the fact that `moduleClass.companion` is `NoSymbol` sometimes is an instance of SI-7567. Comparing the type names directly seems to work:
````
companionResultType.typeSymbol.name.toTypeName == moduleClass.name.toTypeName
````
Perhaps it would be better to use this above then what I current have:
````
moduleClass.companion == companionResultType.typeSymbol
````
What do you think?

Reply:
Okay, I made this change in a second commit. Please have a look.